### PR TITLE
Replace usages of Assembly.CodeBase with Assembly.Location

### LINKED ISSUE #14827: [MVC] Replace usages of Assembly.CodeBase with Assembly.Location
We are using Assembly.CodeBase across the code when we need to find the location of an assembly. This was based on the need to support shadow-copy on full framework.

We spoke with the .NET Core folks and they've essentially told us that Location and CodeBase should be equivalent in .NET Core and that Location is preferred.

Given that we don't support full framework anymore, we should replace all the usages of Assembly.CodeBase across the board in favor of Assembly.Location.

See https://github.com/aspnet/AspNetCore/issues/14501#issuecomment-538606720 for details on the discussion.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/dotnet/aspnetcore/issues/14827

### CODE REVIEW

**Comment by davidfowl** [comment]
We should also handle the location being empty

**Comment by pranavkm** [comment]
> We should also handle the location being empty

MVC already did this. I updated the StaticAssets code to manage it. The remainder of the changes are in test code.

File: src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs [hunk 1]
```diff
@@ -66,7 +66,7 @@ public static IReadOnlyList<Assembly> GetRelatedAssemblies(Assembly assembly, bo
 
             // MVC will specifically look for related parts in the same physical directory as the assembly.
             // No-op if the assembly does not have a location.
-            if (assembly.IsDynamic || string.IsNullOrEmpty(assembly.CodeBase))
+            if (assembly.IsDynamic || string.IsNullOrEmpty(assembly.Location))
```
Comment (javiercn) [inline src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs:69]: Is this right? Will this not fail with the new publish single file work?

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs:69]: I'm not entirely sure how that works, but we should tackle that separately. This doesn't make the code more incorrect than before.

Comment (javiercn) [reply src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs:69]: That's fair, can we file an issue to track that work?

**Review by javiercn — APPROVED** [review]

File: src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs (hunk 1 above)
Comment (davidfowl) [reply src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs:69]: Why check IsDynamic?

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs:69]: We had issues where users had controllers in dynamically generated assemblies.

Comment (davidfowl) [reply src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs:69]: But what does that check have to do with anything here?

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/ApplicationParts/RelatedAssemblyAttribute.cs:69]: * The way you add the controller assembly is by registering it as an application part. 
* MVC will attempt to look up related parts by using the assembly's location. 
* Accessing the location property on a dynamically generated assembly throws.

If the stars align, MVC will throw. Does that help?

**Pull request merged**